### ISSUE TITLE
Is 1MP fixed resizing necessary for the TextEncodeQwenImageEdit node?

### ISSUE THREAD

**Issue by nomadoor** [comfyanonymous/ComfyUI#9481]
### Feature Idea

Currently, the TextEncodeQwenImageEdit node automatically rescales input images so that their total pixel count is fixed to around 1 million.  
Is it technically necessary for this processing step to always rescale inputs to ~1 million pixels?

### Resolution Range
- The base Qwen-Image model works well in the range of 1.5M–17M pixels.
- Qwen-Image-Edit also works properly within this range without particular issues.
- In fact, image quality tends to improve as resolution increases.

**Example**
`prompt : remove the human, keep film grain`
<table>
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/03a03040-48da-488e-a5d0-f5323139050c" width="200"><br>
      input
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/afe4cad4-c817-4b45-b024-8ea1c2cb5211" width="200"><br>
      1024px
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/a8d4691c-d116-4ed4-be43-1276084cbffc" width="200"><br>
      1328px
    </td>
  </tr>
</table>

**Workflow**

<img src="https://github.com/user-attachments/assets/9f297d33-b639-4632-9007-c06e5488bc5e" width="800" alt="Workflow">

[Qwen-Image-Edit_fp8.json](https://github.com/user-attachments/files/21922959/Qwen-Image-Edit_fp8.json)

### User Experience Issues
- Since the TextEncodeQwenImageEdit node forcibly resizes to 1 million pixels (while preserving aspect ratio), when the latent image passed to KSampler is not originally 1M pixels, unintended zooming may occur in the output.
- Users who are not aware of this internal resizing may encounter unexpected zooming, which could be frustrating.

## Summary
- Is there a specific technical reason why Qwen-Image-Edit must be fixed at 1 million pixels?
- Since existing workflow templates already include a **Scale Image to Total Pixels** node, wouldn't it be sufficient to simply explain the recommended resolution in documentation rather than enforcing it within this node?

**Comment by Dolyfin** [comment]
Oh my is this why there is a slight zoom i could not fix? Hidden actions like this should be at least clearly labled and at minimum a toggle.

I have been using a scaler + rounding to multiple of 16 which will conflict and worsen output when I dont manually remove the code.

**Comment by nomadoor** [comment]
@Dolyfin 
The difference in size between the reference latent and the latent image is certainly one major factor, but I think the slight zoom in/out artifacts also come from another mismatch.

Specifically, the VAE compresses the image by a factor of 8, while Qwen2.5-VL uses a patch size of 14 for tokenization. If the resolution is only normalized (e.g. to 1M pixels), these two scales (8 vs. 14) do not align cleanly, which can introduce inconsistencies and subtle zoom effects.

By rounding the resolution to a multiple of the least common multiple of 8 and 14 (which is 56), both sides align properly. In my tests, when I used multiples of 16 or 32, I still observed the unexpected zoom effect, but with multiples of 56 it became much less noticeable (though the model can still sometimes zoom incorrectly for other reasons).

I am not very familiar with these internal mechanisms, so I would definitely welcome corrections from people with more expertise. But at least in my limited experiments, this approach seemed to work.

| Input (divisible by 56) | Output |
|-------------------------|--------|
| ![Input](https://github.com/user-attachments/assets/1878495a-bca1-4358-833a-7bc1644bcd2e) | ![Output](https://github.com/user-attachments/assets/fe1ff6f5-4dda-4059-a415-303a6ed61ead) |

### workflow
<img src="https://github.com/user-attachments/assets/479c5e64-a42b-4dc4-ad25-cfd02ad4437a" width="600" />

[Qwen-Image-Edit_divisible-by-56.json](https://github.com/user-attachments/files/21930665/Qwen-Image-Edit_divisible-by-56.json)

**Comment by Dolyfin** [comment]
This is interesting to hear. I've have tried 16, 32, 64 with various images and would sometimes get almost but not quite perfect and other times very noticeably zoomed outputs. Hoping someone more knowledgeable is able to smooth out the workflow and we get updated templates.

**Comment by lihaoyun6** [comment]
@nomadoor  

<img width="155" alt="Image" src="https://github.com/user-attachments/assets/8861a2f4-2515-4877-8166-403aa16f2a47" />

I found that the Qwen-Image's clip cannot handle image input that deviates too much from 1 megapixels (tested with my custom nodes)  
If you feed the clip an image smaller than 0.3 megapixels or larger than 1.4 megapixels, its output will not work properly.  
The checkpoint can infer oversized canvases, but the clip has this limitation.  
This is why the `TextEncodeQwenImageEdit` node downscale your image.

**Comment by rzgarespo** [comment]
> [@nomadoor](https://github.com/nomadoor)

> I found that the Qwen-Image's clip cannot handle image input that deviates too much from 1 megapixels (tested with my custom nodes) If you feed the clip an image smaller than 0.3 megapixels or larger than 1.4 megapixels, its output will not work properly. The checkpoint can infer oversized canvases, but the clip has this limitation. This is why the `TextEncodeQwenImageEdit` node downscale your image.

oh, I see, not sure but this fix might be related to this issue:
https://www.modelscope.cn/models/DiffSynth-Studio/Qwen-Image-Edit-Lowres-Fix

**Comment by nomadoor** [comment]
@rzgarespo 
> I'm not sure why this is happening to you guys, or I might not have understood the issue, but I don't have those problems with supported and custom sizes.: 784x1232 - 928x1664 and resizing involving 'crop' will inevitably have a zoom-in effect

In your workflow, you are using CLIP Text Encode + ReferenceLatent instead of TextEncodeQwenImageEdit.
However, with Qwen-Image-Edit, it requires not only the latent but also the image information interpreted by Qwen2.5-VL.
Therefore, even if you use a ReferenceLatent node, you would still need to include TextEncodeQwenImageEdit (without feeding it into the VAE).

**Comment by nomadoor** [comment]
@lihaoyun6 
It was a blind spot for me that this limitation comes from Qwen-Image’s CLIP (not being able to handle images that are too small or too large). I see, that makes sense.
However, as @rzgarespo mentioned, there seem to be solutions such as LoRAs that address this issue.

Therefore, I think it would be more flexible and intuitive to handle resizing in a separate node, rather than embedding it directly inside TextEncodeQwenImageEdit.

**Comment by Dolyfin** [comment]
> I'm not sure why this is happening to you guys, or I might not have understood the issue, but I don't have those problems with supported and custom sizes.: 784x1232 - 928x1664 and resizing involving 'crop' will inevitably have a zoom-in effect

The zoom in effect is comparing the already scaled image to final output image. It is not comparing the original image to the cropped 1MP/output image.


Also `Scale Image to Total Pixels` is using a different rounding to `TextEncodeQwenImageEdit`. Scale Image to Total Pixels set to 1MP gives different output resolution than TextEncodeQwenImageEdit. 1333x787 vs 1328x784

After removing the resizing code from `TextEncodeQwenImageEdit` and ensuring that both it and the sampler is getting image of the same resolution, some images still come out zoomed, but manually setting specific resolutions seems to be perfect.

good:
1024x1024, 1184x880, 1392x752, 1568x672

bad:
1104x944, 1248x832, 1328x800, 1456x720, 1504x688

I couldnt really find an official list of resolutions/aspect ratio. The API just has a min and max resolution limit. If someone knows better lmk.

**Comment by nomadoor** [comment]
@Dolyfin 
> After removing the resizing code from `TextEncodeQwenImageEdit` and ensuring that both it and the sampler is getting image of the same resolution, some images still come out zoomed, but manually setting specific resolutions seems to be perfect.

Could you maybe open a pull request with your modified TextEncodeQwenImageEdit node (with the resizing code removed)?

I’m not sure whether this change would be accepted as-is, but I think it would be really helpful to get the ComfyUI developers’ feedback and see what action they might take.

**Comment by tnbyguy** [comment]
I tried removing resizing from TextEncodeQwenImageEdit node and did some testing by resizing source images with a simple resize node before sending them into prompt nodes.

# TLDR
Resizing doesn't seem to matter, as long as it's not too large. Just try different seeds , or maybe crop a square.

First, the logic in current vanilla TextEncodeQwenImageEdit is the same as the ImageScaleToTotalPixels node (the code looks literally the same now), and the outputs are hash identical when I resize to 1m pixels this way in all my tests. It's also identical when using simple resize nodes if aspect ratio fits perfectly and no padding is required.

# Test setup
Prompt nodes are disconnected from VAE, and reference latent nodes are used after prompt nodes. Source images are directly fed into reference latent nodes after VAE encoding. Resized images are only fed into prompt nodes.
Q8_0 weights, fp8 scaled encoder, Euler, Simple, cfg 4, 20 steps, no lightning lora, Pytorch attention.
Resize interpolation: area, unless marked as lanczos

The source images used in testing are cropped from the 1152x2016 forest elf girl example image of Nova Anime3D XL v4.0 on civit. I cropped from bottom up, and the sizes are 1152x1728 (2:3), 1152x1536 (3:4), and 1152x1152 (1:1), all aspect ratios officially supported.
Three random seeds are tested: Seed 1 = 643571810992611, Seed 2 = 1029009223961037, Seed 3 = 123571810992456.
Prompt: Remove her pantyhose. Keep her brown boots and everything else unchanged.

Diffuser pipeline for QwenEdit resizes source images by targeting 1m pixels (1,048,576=1024×1024) but rounding height and width to nearest multiple of 32.
ComfyUI's TextEncodeQwenImageEdit node currently just targets 1m pixels.
These two resizing methods were always tested for each seed and source image. I also tested a few other target resolutions for resizing (marked as additon).

Results marked by X are where model failed to follow prompt or completely fell apart. Generally resolutions > 1.2m pixels end up like this.

# Source 1152x1536 (3:4)

diffuser resize: 896 1184
Comfy resize: 887 1182

Seed 1
additon:
928 1237
960 1280 X
No zoom, except the failed one.

Seed 2
additon:
864 1152
928 1237
960 1280 X
All zoom.

Seed 3
additon:
864 1152
960 1280 X ()
All zoom, except the failed one which somehow didn't zoom lol.

# Source 1152x1728 (2:3)

diffuser resize: 832 1248
Comfy resize: 836 1254

Seed 1
additon:
832 1248 lanczos
836 1254 lanczos
All zoom.

Seed 2
additon:
832 1248 lanczos
896 1344
928 1392 X
960 1440 X
No zoom. Even the failed ones.

Seed 3
additon:
896 1344
All zoom.

# Source 1152x1152 (1:1)

diffuser resize: 1024 1024
Comfy resize: 1024 1024

Seed 1 & 3
Only tested with 1024x1024 comfy resize. No zoom.

Seed 2
1064 1064
1094 1094 lanczos X (almost okay)
1094 1094 X
1152 1152 X
No Zoom. Even the failed ones.

# Extra notes:
When the output is zoomed, every resizing method is just a little different how much it zooms.

**Comment by zhzLuke96** [comment]
This is a very good discussion. I also think the **TextEncodeQwenImageEdit** node does have issues, which seriously underestimates the capability of the Qwen-Image-Edit model. In some specific domains, I even find this model more useful than the Nano-Banana model.

Regarding the **1MP limitation**, I believe it is indeed necessary, because this is the value officially recommended. From my own tests, resolutions outside of 1MP tend to shift the output behavior more towards **“generating new images”** rather than **“editing existing ones.”**

I wrote a custom node to help test and address this issue:
👉 [https://github.com/lenML/comfyui\_qwen\_image\_edit\_adv](https://github.com/lenML/comfyui_qwen_image_edit_adv)

In my opinion, the pixel shift / incorrect scaling mainly comes from three factors:

1. As mentioned earlier, **VAE encoding requires dimensions divisible by 8**.
2. Inside **TextEncodeQwenImageEdit**, the node calls the VAE to generate `ref_latent`. When `vae.encode` is invoked, alignment logic is still applied, which may cause mismatches between `ref_latent` and the final target resolution.
3. The **1MP constraint**, which I suspect is related to the training data. The exact reason is unclear.

As for @nomadoor’s point about aligning with CLIP’s **14×14 patch size**, I don’t think that’s strictly necessary. My understanding is that the CLIP model compresses the input image into a semantic vector, and therefore it shouldn’t directly impact pixel-level alignment.

However, even if we fully follow these constraints, **shifts can still occur**. I suspect this final offset problem comes from the *editing* training stage itself, where pixel alignment may not have been emphasized. Because of this, I suggest adding an explicit consistency requirement in the prompt, for example:
**“Keep the composition and aspect ratio unchanged.”**

---

Additional note:
The official recommended scaling alignment for Qwen-Image is **32**.

Reference:
[https://github.com/huggingface/diffusers/blob/d06750a5fd19781de68066bb34a3520af83cf124/src/diffusers/pipelines/qwenimage/pipeline\_qwenimage\_edit.py#L155-L162](https://github.com/huggingface/diffusers/blob/d06750a5fd19781de68066bb34a3520af83cf124/src/diffusers/pipelines/qwenimage/pipeline_qwenimage_edit.py#L155-L162)

**Comment by nomadoor** [comment]
@zhzLuke96 
Thank you very much for developing the custom node and for conducting such thorough testing.

I was also skeptical about the CLIP patch size argument. While it may not be completely irrelevant, I agree that the major causes of the zoom effect are more likely due to the VAE and the inherent limitations of the model itself.
If we truly aim for pixel-perfect alignment, I feel that a new approach, such as pixel diffusion models, might be necessary.

In any case, I really appreciate your work—thank you!
